kernel/libs/aster-bigtcp/src/iface/common.rs

```diff
@@ -109,47 +168,43 @@ impl<E> IfaceCommon<E> {
         }
     }
 
-    pub(super) fn bind_socket(
-        &self,
-        iface: Arc<dyn Iface<E>>,
-        socket: Box<AnyUnboundSocket>,
-        config: BindPortConfig,
-    ) -> core::result::Result<AnyBoundSocket<E>, (BindError, Box<AnyUnboundSocket>)> {
-        let port = if let Some(port) = config.port() {
-            port
-        } else {
-            match self.alloc_ephemeral_port() {
-                Some(port) => port,
-                None => return Err((BindError::Exhausted, socket)),
-            }
-        };
-        if !self.bind_port(port, config.can_reuse()) {
-            return Err((BindError::InUse, socket));
-        }
+    /// Removes a socket from the interface.
+    pub(crate) fn remove_socket(&self, handle: SocketHandle) {
+        self.sockets.lock().remove(handle);
+    }
 
-        let (handle, socket_family, observer) = match socket.into_raw() {
-            (AnyRawSocket::Tcp(tcp_socket), observer) => (
-                self.sockets.lock().add(tcp_socket),
-                SocketFamily::Tcp,
-                observer,
-            ),
-            (AnyRawSocket::Udp(udp_socket), observer) => (
-                self.sockets.lock().add(udp_socket),
-                SocketFamily::Udp,
-                observer,
-            ),
-        };
-        let bound_socket = AnyBoundSocket::new(iface, handle, port, socket_family, observer);
-        self.insert_bound_socket(bound_socket.inner());
+    pub(crate) fn remove_bound_socket_now(&self, socket: &Arc<AnyBoundSocketInner<E>>) {
+        let keyable_socket = KeyableArc::from(socket.clone());
 
-        Ok(bound_socket)
+        let removed = self
+            .bound_sockets
+            .write_irq_disabled()
+            .remove(&keyable_socket);
+        assert!(removed);
     }
 
-    /// Remove a socket from the interface
-    pub(crate) fn remove_socket(&self, handle: SocketHandle) {
-        self.sockets.lock().remove(handle);
+    pub(crate) fn remove_bound_socket_when_closed(&self, socket: &Arc<AnyBoundSocketInner<E>>) {
```

Comment:
This two methods `remove_bound_socket_now` and `remove_bound_socket_when_closed` share the same code for removing socket from the bound socket, so they can be simplified.

```rust
fn remove_bound_socket_now(&self, socket: &Arc<AnyBoundSocketInner<E>>) -> KeyableArc<AnyBoundSocketInner<E> {
    let keyable_socket = KeyableArc::from(socket.clone());

    let removed = self
            .bound_sockets
            .write_irq_disabled()
            .remove(&keyable_socket);
    assert!(removed);
   keyable_socket
}

fn remove_bound_socket_when_closed(&self, socket: &Arc<AnyBoundSocketInner<E>>) {
    let keyable_socket = remove_bound_socket_now();
    //......
}
```

And how do you like renaming `remove_bound_socket_when_closed` as `remove_bound_socket_delayed`, because I found `remove_bound_socket_now` is also called when closing socket.

Reply:
Yes, I think your naming is better than the current one.

However, all of the things mentioned in this comment are _not_ introduced in this PR, and more importantly, the related logic will soon be refactored in the Ethernet/IP protocol implementation and packet dispatching, so I don't think it needs to be revised. I'm afraid the revision will do nothing but introduce some conflicts that need to be resolved.

Reply:
Make sense. Then we can just keep the current from in this PR.